File: JavaPartI/la_middleNodeOfLinkedList.java
/**
 * class ListNode {
 *   public int value;
 *   public ListNode next;
 *   public ListNode(int value) {
 *     this.value = value;
 *     next = null;
 *   }
 * }
 */
public class Solution {
  public ListNode middleNode(ListNode head) {
    // write your solution here
    ListNode slow = head, fast = head;
    if (head == null) {
      return head;
    }
    while ( fast.next!= null && fast.next.next!=null) {
      slow = slow.next;
      fast = fast.next.next;
    }
    return slow;
  }
}

File: pythonPart/la_middleNodeOfLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution(object):
  def middleNode(self, head):
    """
    input: ListNode head
    return: ListNode
    """
    # write your solution here
    if head == None:
      return head
    slow = head
    fast = head
    while fast.next != None and fast.next.next != None:
      fast = fast.next.next
      slow = slow.next
    return slow
